### ISSUE TITLE
+disconnect clauses are computed multiple times. 

### ISSUE THREAD

**Issue by b-scholz** [souffle-lang/souffle#1176]
Soufflé transforms the input program
```
.decl A(x:number)
A(1).
.decl B(x:number)
B(2).
.decl C(x:number)
.printsize C
C(1) :- A(x), B(x).
C(2) :- A(x), B(x).
```
to
```
.decl +disconnected0()  
+disconnected0() :- 
   A(x),
   B(x).
.decl +disconnected1()  
+disconnected1() :- 
   A(x),
   B(x).
.decl A(x:number)  
A(1).
.decl B(x:number)  
B(2).
.decl C(x:number)  
C(1) :- 
   +disconnected0().
C(2) :- 
   +disconnected1().
.printsize C(IO="stdoutprintsize")
```
which is sub-optimal since the cross-product ```A(x), B(x)``` 
is computed twice.

**Comment by azreika** [comment]
Since partitioning is a clause-local transformation, this issue actually falls under having a stronger program-minimising transformation (`MinimiseProgramTransformer.cpp`). Currently, that transformation is only relation-local, but should be extended to find equivalent (or subsuming) relations, too, which would solve this particular problem.

**Comment by azreika** [comment]
Since this pattern of equivalent singleton relations appears very often with the `PartitionBodyLiteralsTransformer`, I think a singleton-relation-specific extension to the program-minimising transformation would be very appropriate. I will pull up an extension soon.

**Comment by b-scholz** [comment]
I agree - we only need to consider the equivalence of two singleton-relations. Perhaps we exhaustively try to match or build some notions of a normal form. 

We need to be careful with output relations, i.e., we should not merge them (or have a simple copy rule instead) so that the output of the program is not disturbed. 

Note that this may be a recursive procedure. I.e., after finding one matching pair, more matching pairs may be found after coalescing them.

**Comment by azreika** [comment]
This case was resolved by PR #1179.
